### DISCUSSION TITLE
Proxy Yield Aggregator v1.0.0 - ERC4626 Compatible Deposit Contract

### DISCUSSION THREAD

**Discussion by 0x4007** [pavlovcik/uad-ui#602] [Protocol Specifications]
### Discussed in https://github.com/ubiquity/ubiquity-dollar-development/discussions/171

<div type='discussions-op-text'>

<sup>Originally posted by **pavlovcik** May 14, 2022</sup>
# "Bring Your Own Yield Source"

For the next iteration of the Proxy Yield Aggregator (PYA) program a user should be able to bring their own [ERC4626](https://github.com/fei-protocol/ERC4626/blob/7a947f2507b760ae470578cfb106f71ff5b1a14b/src/ERC4626Router.sol) compatible yield source, and connect it to our deposit contract. 

This product is intended for users to be able to stack the total yield that they can receive from their favorite yield bearing sources. 

Upon withdrawal, the user interface should present a slider which represents risk tolerance of the user. 

## Example

1. The user likes a new ERC4626 Compatible Yearn USDC vault that promises 20% APR.
1. The user "connects" the ERC4626 vault to our deposit contract. 
    1. Implementation details must be researched, but I hope that we can avoid a contract redeployment, a la deploying a new Uniswap market.
1. The user deposits 1000 USDC, and the USDC is immediately transferred to the underlying yield source.
1. One year later, the user seeks to withdraw their 1200 USDC, and returns to our user interface to withdraw.
1. The user interface presents a slider, which seeks to understand the user's risk/reward tolerance 
    1. For example, the user can withdraw with no additional risk (1200 USDC) or the user can withdraw with max risk (2400 uCR)
    1. The idea is that the user can convert the USDC into our uCR for a premium 
    1. In exchange, the Ubiquity Dollar is able to raise collateral for the dollar in perpetuity.
1. User withdraws at "25% r/r" netting them something like 900 USDC and 900 uCR (premium curve formula to be determined)
    1. "50% r/r" would be 600 USDC and 1200 uCR
    1. "100% r/r" is 0 USDC and 2400 uCR
    1. "16.7% r/r" is 1000 USDC and 400 uCR // I imagine that many would, in this example go here or lower in order to only convert yield, and to receive the principle back.

## Adjustable Parameters

1. uCR Premium, a math formula (default would be a linear calculation, 1 USDC -> 2 uCR)
    1. This would be deployed as a separate smart contract
    1. This gives us flexibility to implement exponential or logarithmic functions which may more efficiently produce our desired economic outcomes!
1. Yield token whitelist (default would be an array with the following contract addresses: USDC, DAI, LUSD)
    1. The idea is that, because the user is able to bring their own yield source, we wouldn't want a situation where they give assets that we do not want (like UST)
    1. Rejection should happen upon deposit, not withdrawal; in case we change our accepted assets after the user already has deposited their tokens.

## Remarks

It would be very interesting to match the uCR Premium (math formula) with the bond of the associated asset. That way we are purchasing at the same price premiums ecosystem wide. As long as we're mindful of 1. the interface implementation, and 2. that we architect it so we can associate the uCR Premium PER asset, then I think we can easily share the same formula between our bonds and the PYA. </div>

**Comment by LogETH** [comment]
This bounty seems to be asking to simply fork the ERC4626 router contract, add an option to withdraw in an alternate token with a custom reward curve and restricting the router to only certain stablecoins, that seems too easy for a task with that tall of a bounty, is there anything I'm missing? 

In fact, with the implementation I'm thinking of, you don't even have to deploy new vault contracts, you could just reuse already existing vault contracts that are already deployed, audited, and used on chain

**Comment by 0x4007** [comment]
> This bounty seems to be asking to simply fork the ERC4626 router contract, add an option to withdraw in an alternate token with a custom reward curve and restricting the router to only certain stablecoins, that seems too easy for a task with that tall of a bounty, is there anything I'm missing?
> 
> In fact, with the implementation I'm thinking of, you don't even have to deploy new vault contracts, you could just reuse already existing vault contracts that are already deployed, audited, and used on chain

1. This isn't assigned to @ubiquity-bounties so this isn't an open bounty. 
2. You can check the associated discussion for more details and q&a on the specification, but the point of the idea is to stack yield in our uCR token on another yield source, and use that yield to further collateralize our dollar; so in your description I don't see how that could be possible.

**Comment by LogETH** [comment]
I see, thank you!

**Comment by codeawayaks** [comment]
Please assign this to me.

**Comment by sergfeldman** [comment]
> Please assign this to me.

@codeawayaks Hi. As you can see in the assignee and comments on this issue, it is not available. 

The first time you wrote requests in more than 20 tasks, and after that, you didn’t answer which one you wanted to take.
This is the second time you've posted an irrelevant request for the bounty.

Unfortunately, this looks like spam, so we will block your ability to comment for the upcoming 30 days.

**Comment by 0x4007** [comment]
Hey @zgorizzo69 looks like you were assigned this a long time ago. What's the status?

**Comment by zgorizzo69** [comment]
Proxy yield has been merged inside #267 I believe. It is not part of the dev branch anymore you said it is because it is being replaced by the new one. Can you explain exactly what the new version is about ?

**Comment by 0x4007** [comment]
Then there was some miscommunication somewhere. As you know, you built an earlier version of the pya. That is deprecated for the one you delivered in #267 which is ERC4626 compliant. If the ERC4626 compliant pya has been deleted then it should be restored.

I remember now that we need to be able to manage all of deployed pyas and we decided to hold off until the Diamond Ubiquity Dollar Manager contract is complete so that we can add a pya/vault manager facet right?

**Comment by zgorizzo69** [comment]
I can retrieve my version and make it a facet yes

**Comment by 0x4007** [comment]
To clarify, I remember our conversation leading to us splitting the deliverable for this into two contracts.

1. is a factory that allows users to deposit (through us) to their favorite vault strategy. The way that this works is similar to making a new Uniswap market with two assets that you choose. But instead, the user chooses what vault they want to connect to.
2. is a manager for all global settings (e.g. fee and uCR boost) of these deployed deposit contracts

The manager is to reside on the Diamond, right?

**Comment by 0x4007** [comment]
Also maybe useful to add this to our tests https://github.com/a16z/erc4626-tests

**Comment by zgorizzo69** [comment]
> To clarify, I remember our conversation leading to us splitting the deliverable for this into two contracts.
> 
>     1. is a factory that allows users to deposit (through us) to their favorite vault strategy. The way that this works is similar to making a new Uniswap market with two assets that you choose. But instead, the user chooses what vault they want to connect to.
> 
>     2. is a manager for all global settings (e.g. fee and uCR boost) of these deployed deposit contracts
> 
> 
> The manager is to reside on the Diamond, right?

yes

**Comment by 0x4007** [comment]
looks like there's a security issue around being the first depositor 

> That’s right, the price for 1 wei of shares becomes 1e18 of assetToken! Which means that with the rounding down of the division, any following user depositing less than 1e18 of assetTokens will receive 0 shares and lose their funds!
4/7

https://twitter.com/cergyk1337/status/1612510626517241857?s=46&t=eahphooxBpk3Lfkdrt9KTQ

https://github.com/sherlock-audit/2022-10-astaria-judging/issues/143

https://github.com/code-423n4/2022-04-xtribe-findings/issues/66

https://github.com/sherlock-audit/2022-11-sense-judging/issues/50

**Comment by zgorizzo69** [comment]
damn better fix that asap !

**Comment by 0x4007** [comment]
Yearn released permission-less vaults. We can probably learn something from their code. 

https://twitter.com/iearnfinance/status/1612484330755919880?s=46&t=eahphooxBpk3Lfkdrt9KTQ

**Comment by nshmadhani** [comment]
is the bounty still open?

**Comment by sergfeldman** [comment]
> is the bounty still open?

@nshmadhani no, this bounty is assigned.

**Comment by 0x4007** [comment]
@zgorizzo69 what's the status on it?
